### PR TITLE
Mature Price Models for "v1" ENSAnalytics

### PR DESCRIPTION
**Pull request by Goader**
# ENS Referrals API v1: Price Types Migration

closes: #1521

---

## Reviewer Focus (Read This First)

<details>
<summary>What reviewers should focus on</summary>

- New v1 API implementation in `packages/ens-referrals/src/v1/*`
- `/v1` subpath export
- Package configuration
- Calculation of `awardPoolApproxAmount`

Error I got during testing:

**Location**: `apps/ensapi/src/handlers/ensanalytics-api-v1.test.ts`

**Error**: Tests fail with:
```json
{
  "responseCode": "error",
  "error": "Internal server error",
  "errorMessage": "(0 , __vite_ssr_import_0__.serializePriceUsdc) is not a function"
}
```

In my various experiments, I found it to be also failing with `serializePriceEth`, which was implemented before this PR. Tried untangling imports in `ensnode-sdk/ens` and `ensnode-sdk/shared` to make this dependence a single-direction one, but the problem persisted. Importing this directly from ENSApi seemed to work, and importing it from the `dist` version worked as well. The current workaround is to export it from `internal` and use it for importing in `ens-referrals`, which is not what I intended, but from the fixes I found seemed to be the most benign.

Fixed through: #1572

</details>

---

## Problem & Motivation

<details>
<summary>Why this exists</summary>

* When the `ens-referrals` package was first implemented, we took a few temporary shortcuts to ensure ENS Holiday Awards could be shipped before Dec 1. These temporary shortcuts included quick fixes for the `USDQuantity` and `RevenueContribution` price data models currently defined in `ens-referrals`.
* A more mature price data model for `PriceUsdc` and `PriceEth` are defined in `ensnode-sdk`, however we were previously blocked on using these more mature data models because of [issue 1519](https://github.com/namehash/ensnode/issues/1519) and our need to ensure ENS Holiday Awards was shipped on schedule.
* See issue: #1521

</details>

---

## What Changed (Concrete)

<details>
<summary>What actually changed</summary>

1. Copy-pasted the whole `ens-referrals` to `v1` directory
2. Added appropriate entry point for `v1` and `v1/internal`
3. Updated `USDQuantity` and `RevenueContribution` types to use mature price types from `ensnode-sdk`
4. Added `V1` middleware to ENSApi to properly handle new types that appear in Referrer Leaderboard
5. Changed imports from `ensnode-sdk/shared` to be explicit to avoid Vitest issues
6. Changed the endpoint path from `/ensanalytics/v1` to `/v1/ensanalytics`

</details>

---

## Design & Planning

<details>
<summary>How this approach was chosen</summary>

- See issue #1521
- Tricky part was versioning of impacted models. I chose to separate them at the file level, and added a separate import path, so that ENSApi can be easily migrated to use the updated versions.
- Essentially copy-pasted the whole package into `v1/` version

</details>

- Planning artifacts: see issue #1521
- Copy-pasted files, updated impacted fields.
- Reviewed / approved by: N/A

---

## Self-Review

<details>
<summary>What you caught yourself</summary>

- Using old middleware, which produced an old version of referrer leaderboard
- Lack of proper award pool shared calculation
- Old logging of leaderboard building was trying to build JSON out of rules, that now need explicit serialization, because of `bigint`
- Referrer Detail function in client was using incorrect endpoint path starting with `api`. It was a bug introduced previously in `ensnode-sdk` and migrated to `ens-referrals`. Fixed as of now

</details>

- Bugs caught: Vite SSR (fixed in #1572)
- Logic simplified: None
- Naming / terminology improved: None
- Dead or unnecessary code removed: None

---

## Cross-Codebase Alignment

<details>
<summary>Related code you checked</summary>

- Verified v0 API remains completely unchanged
- Checked `PriceEth` and its serialization implementation for consistency

</details>

- Search terms used: `PriceUsdc`, `PriceEth`, `awardPoolApproxValue`, `totalAwardPoolValue`, `RevenueContribution`, `USDQuantity`
- Reviewed but unchanged: v0 API implementation (still uses `USDQuantity` and `RevenueContribution`)
- Deferred alignment: Some of the issues outlined by CodeRabbit in comments (zero division if `maxQualifiedReferrers` is 1, using set for `uniqueReferrers`)

---

## Downstream & Consumer Impact

<details>
<summary>Who this affects and how</summary>

- Shouldn't impact anything apart from v1 endpoints

</details>

- Public APIs affected: v1 API only
- Docs updated: Copied from v0
- Naming decisions: used `v1` path for imports

---

## Testing Evidence

<details>
<summary>How this was validated</summary>

- Standard CI testing
- Manual validation

</details>

- Testing performed: Automated testing + manual validation
- Known gaps: None
- What reviewers have to reason about manually: `v1` for importing, Vite issues, proper currency calculations

---

## Scope Reductions

<details>
<summary>What you intentionally didn't do</summary>

No changes to v0 implementation.

</details>

- Follow-ups: None
- Why they were deferred: N/A

---

## Risk Analysis

<details>
<summary>How this could go wrong</summary>

Low risk - v0 API completely unchanged, v1 is additive only.

</details>

- Risk areas: None (v0 unchanged, v1 additive)
- Mitigations or rollback options: Trivial to revert, nothing depends on v1 yet
- Named owner if this causes problems: @Goader

---

## Pre-Review Checklist (Blocking)

- [x] I reviewed every line of this diff and understand it end-to-end
- [x] I'm prepared to defend this PR line-by-line in review
- [x] I'm comfortable being the on-call owner for this change
- [x] Relevant changesets are included (or explicitly not required)

### CODE REVIEW

File: packages/ens-referrals/src/v1/api/serialized-types.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import type { AggregatedReferrerMetrics } from "../aggregations-v1";
+import type { ReferrerLeaderboardPage } from "../leaderboard-page-v1";
+import type { ReferrerDetailRanked, ReferrerDetailUnranked } from "../referrer-detail-v1";
+import type { AwardedReferrerMetrics, UnrankedReferrerMetrics } from "../referrer-metrics-v1";
+import type { ReferralProgramRules } from "../rules-v1";
+import type {
+  ReferrerDetailResponse,
+  ReferrerDetailResponseError,
+  ReferrerDetailResponseOk,
+  ReferrerLeaderboardPageResponse,
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply packages/ens-referrals/src/v1/api/serialized-types.ts]: @Goader Could you check this?

File: packages/ens-referrals/src/v1/api/zod-schemas.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    responseCode: z.literal(ReferrerDetailResponseCodes.Error),
+    error: z.string(),
+    errorMessage: z.string(),
+  });
+
+/**
+ * Schema for {@link ReferrerDetailResponse}
+ */
+export const makeReferrerDetailResponseSchema = (valueLabel: string = "ReferrerDetailResponse") =>
+  z.union([
```
Comment (lightwalker-eth) [inline packages/ens-referrals/src/v1/api/zod-schemas.ts]: I believe here we could make use of Zod's support for discriminated unions.

`discriminatedUnion` instead of `union`.

Related docs: https://zod.dev/api?id=discriminated-unions

File: packages/ens-referrals/src/v1/link.test.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { type Address, getAddress } from "viem";
+import { describe, expect, it } from "vitest";
+
+import { buildEnsReferralUrl } from "./link";
+
+const vitalikEthAddressLowercase: Address = "0xd8da6bf26964af9d7eed9e03e53415d37aa96045";
+const vitalikEthAddressChecksummed: Address = getAddress(
+  "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
+);
```
Comment (lightwalker-eth) [inline packages/ens-referrals/src/v1/link.test.ts]:
```suggestion
const vitalikEthAddressChecksummed: Address = getAddress(vitalikEthAddressLowercase);
```

File: packages/ens-referrals/src/v1/rules.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
+import {
+  type AccountId,
+  type PriceUsdc,
+  priceUsdc,
+  type UnixTimestamp,
+} from "@ensnode/ensnode-sdk";
+import { makePriceUsdcSchema } from "@ensnode/ensnode-sdk/internal";
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply packages/ens-referrals/src/v1/rules.ts]: @Goader This seems like a good suggestion.

File: packages/ens-referrals/src/v1/leaderboard.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  accurateAsOf: UnixTimestamp;
+}
+
+export const buildReferrerLeaderboard = (
+  allReferrers: ReferrerMetrics[],
+  rules: ReferralProgramRules,
+  accurateAsOf: UnixTimestamp,
+): ReferrerLeaderboard => {
+  const uniqueReferrers = allReferrers.map((referrer) => referrer.referrer);
+  if (uniqueReferrers.length !== allReferrers.length) {
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply packages/ens-referrals/src/v1/leaderboard.ts]: @Goader This looks like a good suggestion.

File: .changeset/thick-mails-appear.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+---
+"@namehash/ens-referrals": minor
+"@ensnode/ensnode-sdk": minor
+"ensapi": patch
+---
+
+Migrated v1 referrer leaderboard API to use mature `PriceEth` and `PriceUsdc` types from `ensnode-sdk`, replacing temporary `RevenueContribution` and `USDQuantity` types. Added `/v1` subpath export to `ens-referrals`.
```
Comment (lightwalker-eth) [inline .changeset/thick-mails-appear.md]: @Goader Could you please confirm that `RevenueContribution` and `USDQuantity` are completely removed for the "v1" code? Asking because I saw a reference to `RevenueContribution` in the "v1" code. Thanks!

File: packages/ens-referrals/src/v1/aggregations.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  validateReferrerScore(metrics.grandTotalQualifiedReferrersFinalScore);
+  validateReferrerScore(metrics.minFinalScoreToQualify);
+};
+
+export const buildAggregatedReferrerMetrics = (
+  referrers: RankedReferrerMetrics[],
+  rules: ReferralProgramRules,
+): AggregatedReferrerMetrics => {
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply packages/ens-referrals/src/v1/aggregations.ts]: @Goader This seems like valid feedback

File: packages/ens-referrals/src/v1/referrer-metrics.ts [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+};
+
+export const buildAwardedReferrerMetrics = (
+  referrer: RankedReferrerMetrics,
+  aggregatedMetrics: AggregatedReferrerMetrics,
+  rules: ReferralProgramRules,
+): AwardedReferrerMetrics => {
+  const awardPoolShare = calcReferrerAwardPoolShare(referrer, aggregatedMetrics, rules);
+
+  // Calculate the approximate USDC value by multiplying the share by the total award pool value
```
Comment (lightwalker-eth) [inline packages/ens-referrals/src/v1/referrer-metrics.ts]: A few suggestions here:

1. Define a utility function that can take any bigint value and scale it by a number. Ex: `scaleBigint`.
2. Write good unit tests for this function and ensure all edge cases are handled.
3. Define a specialized helper for the generic utility function described in point 1. above that is specialized for Price objects. Ex: `scalePrice`.
4. Make use of `scalePrice` here.

File: packages/ens-referrals/src/v1/referrer-metrics.ts [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+  const currencyInfo = getCurrencyInfo(rules.totalAwardPoolValue.currency);
+  const precisionScale = 10n ** BigInt(currencyInfo.decimals);
+  const scaledShare = BigInt(Math.floor(awardPoolShare * Number(precisionScale)));
+  const awardPoolApproxAmount = (rules.totalAwardPoolValue.amount * scaledShare) / precisionScale;
```
[the remark this answers is not in the document]
Comment (lightwalker-eth) [reply packages/ens-referrals/src/v1/referrer-metrics.ts]: @Goader This suggestion is related to my other comment on this file.

File: packages/ens-referrals/src/v1/referrer-metrics.ts [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+ * Build an unranked zero-score referrer record for a referrer address that is not in the leaderboard.
+ *
+ * This is useful when you want to return a referrer record for an address that has no referrals
+ * and is not qualified for the leaderboard.
+ *
+ * @param referrer - The referrer address
+ * @returns An {@link UnrankedReferrerMetrics} with zero values for all metrics and null rank
+ */
+export const buildUnrankedReferrerMetrics = (referrer: Address): UnrankedReferrerMetrics => {
+  const baseMetrics = buildReferrerMetrics(referrer, 0, 0, {
```
Comment (lightwalker-eth) [inline packages/ens-referrals/src/v1/referrer-metrics.ts]: Let's not create Price objects inline like this. Suggest instead to define utility functions such as `buildPriceEth` that take in an amount (in wei) and then return back a `PriceEth` object.

File: apps/ensapi/src/lib/ensanalytics/referrer-leaderboard/mocks-v1.ts [hunk 11]
```diff
@@ -0,0 +1,1540 @@
...
+    minFinalScoreToQualify: 0,
+  },
+  referrers: new Map([
+    [
+      "0x538e35b2888ed5bc58cf2825d76cf6265aa4e31e",
+      {
+        referrer: "0x538e35b2888ed5bc58cf2825d76cf6265aa4e31e",
+        totalReferrals: 3,
+        totalIncrementalDuration: 94694400,
+        totalRevenueContribution: {
```
Comment (lightwalker-eth) [inline apps/ensapi/src/lib/ensanalytics/referrer-leaderboard/mocks-v1.ts]: This can be implemented as `parseEth`. Please see my related comment on `parseUsdc`.
